### PR TITLE
Added form for projects, added views for new&edit project

### PR DESCRIPTION
**Pull request by livike**
Added form for edit and add new projects, using ModelForm.
For now no styling to the form elements.
Related to #121

### CODE REVIEW

**Comment by goranche** [comment]
>  no styling to the form elements

high five!!! :hand:

File: web/views.py [hunk 1]
```diff
@@ -1,6 +1,9 @@
 from django.template import RequestContext
 from django.shortcuts import render_to_response, render, get_object_or_404
+from django.shortcuts import redirect
```
Comment (gandalfar) [inline web/views.py]: this should go one line up

**Comment by gandalfar** [comment]
`project_new` will fail if form is not valid :)

**Comment by anuschka** [comment]
@livike It works great :+1:  other than when entering a new project with status ID in progress. In that case you get a message " Ensure this value has at most 8 characters (it has 10)." Also (but less important) it would be nice to have a description for the  date format dd/mm/yyyy hh:mm :v:

File: web/views.py [hunk 2]
```diff
@@ -1,8 +1,14 @@
 from django.template import RequestContext
 from django.shortcuts import render_to_response, render, get_object_or_404, redirect
+<<<<<<< HEAD
```
Comment (ialja) [inline web/views.py]: Ahem, you know what needs to be done here, yes? :wink:

**Comment by ialja** [comment]
A few suggestions for improvements:

- When adding a new project, there is no info about which fields are mandatory (until you hit Save and get errors). 
- It's not clear which field the error refers to ("This field is required" - which field, the one above or below?)
- Entering project deadline is annoying, especially with no instructions regarding the required format. Perhaps add an interactive [DateTimePicker](http://xdsoft.net/jqplugins/datetimepicker/) for easier input?
- When editing projects, the deadline date and time is not shown in the form.

File: web/views.py (hunk 2 above)
Comment (livike) [inline web/views.py]: Riiight... :blush:
fixed...

File: web/templates/pages/project_edit.html [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{% extends 'base.html' %}
+
+{% block content %}
+	<h1>New project</h1>
+	<form method="POST">{% csrf_token %}
+		{{ projectform.as_p }}
+		<button type="submit" class="save btn btn-default">Save</button>
+	</form>
+{% endblock %}
```
Comment (goranche) [inline web/templates/pages/project_edit.html]: missing newline

File: web/form.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+from django import forms
+
+from api.models import Project
+
+class ProjectForm(forms.ModelForm):
+
+    class Meta:
```
Comment (goranche) [inline web/form.py]: @livike we use tabs for indentation... fix throughout whole file!

File: web/form.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+        model = Project
+        fields = ('project_name', 'project_description','status_id','project_deadline',
```
Comment (ercchy) [inline web/form.py]: This is not very nice to read, how about putting every attribute to new line?

File: web/templates/pages/project_detail.html [hunk 6]
```diff
@@ -3,6 +3,8 @@
 {% block content%}
 	<div class="row">
 		<h3><a href="{% url 'web.views.project_detail' project_id=project.project_id %}">{{ project.project_name }}</a></h3>
+		<a class="btn btn-default" href="{% url 'web.views.project_new' %}" class="top-menu"><span class="glyphicon glyphicon-plus"></span></a>
```
Comment (ercchy) [inline web/templates/pages/project_detail.html]: You did not name your URL and now you have this oogly peace of :hankey:  here. Please name the URL, and change this.

File: web/urls.py [hunk 7]
```diff
@@ -16,5 +16,7 @@
 	url(r'^edituser/$', views.edituser, name='pages-edituser'),
 	url(r'^privacy/$', views.privacy, name='pages-privacy'),
 	url(r'^project/(?P<project_id>[0-9]+)/$', views.project_detail),
```
Comment (ercchy) [inline web/urls.py]: What is going on here? Why are some URLS named and others  aren't? That is some sloppy work!

File: web/views.py [hunk 8]
```diff
@@ -40,7 +42,29 @@ def project_detail(request,project_id):
...
+def project_new(request):
+	if request.method == "POST":
+		projectform = ProjectForm (request.POST)
+		if projectform.is_valid():
+			project = projectform.save()
+			return redirect('web.views.project_detail', project_id=project.project_id)
+	else:
+		projectform = ProjectForm()
+
+	return render(request, 'pages/project_edit.html', {'projectform': projectform})
```
Comment (ercchy) [inline web/views.py]: Ialwaysliketoreadeverythingwithoutspacesoranyothersignbetweewords!!!

File: web/views.py [hunk 9]
```diff
@@ -40,7 +42,29 @@ def project_detail(request,project_id):
...
+
+def project_edit(request, project_id):
+	project = get_object_or_404(Project, project_id=project_id)
```
Comment (ercchy) [inline web/views.py]: Like I said in another PR, you guys are raising 404 error, and you have no template to show when this happens. If there is no template, then one, if in development mode, gets a really ugly error or if production mode everything just falls apart spectacularly!

It is much better to get an empty dict (which the model will return if there will be not projects) and test for that in the template than this.

File: web/views.py (hunk 8 above)
Comment (ercchy) [inline web/views.py]: Why is this rendering `project_edit` template if it is `project_new` function?

You could very elegantly use one method for both creating and editing project.

File: web/views.py [hunk 10]
```diff
@@ -40,7 +42,29 @@ def project_detail(request,project_id):
 	project = get_object_or_404(Project, project_id=project_id)
 	return render(request, 'pages/project_detail.html', {'project': project})
 
+def project_new(request):
+	if request.method == "POST":
+		projectform = ProjectForm (request.POST)
+		if projectform.is_valid():
+			project = projectform.save()
+			return redirect('web.views.project_detail', project_id=project.project_id)
```
Comment (ercchy) [inline web/views.py]: another unnamed url :rolls_her_eyes_:

File: web/views.py [hunk 11]
```diff
@@ -40,7 +42,29 @@ def project_detail(request,project_id):
...
+	else:
+		projectform = ProjectForm(instance=project)
+
+	return render(request, 'pages/project_edit.html', {'projectform': projectform})
+
 def logout(request):
     auth_logout(request)
     return redirect('/')
-
```
Comment (ercchy) [inline web/views.py]: Why did you delete a new line at the end?

Ah I know, to annoy @goranche.... 

Please do add new line to the end of the file.

**Comment by livike** [comment]
@ercchy : made fixes regarding your comments. The template update is coming.

File: web/views.py (hunk 8 above)
Comment (livike) [inline web/views.py]: I was following only the Djangogirls tutorials solution, where they have two methods

**Comment by livike** [comment]
@ialja , thanks for your above suggestions for improvements. As @gandalfar suggested to me, I'm working on the additional and improvement stuff on a new branch based on the projectform branch.

File: web/views.py (hunk 9 above)
Comment (livike) [inline web/views.py]: Added templates for handling 404 and 500 errors

Comment (ercchy) [inline web/views.py]: My suggestion was actually to stop using 404 and retrieve objects via normal Django query https://docs.djangoproject.com/en/1.7/topics/db/queries/#retrieving-objects

**Comment by ercchy** [comment]
@livike I made a screenshot with some more feedback for you.

I suggest you do this one really perfectly, so everyone can use then this as a model

That means some UI and UX thinking is needed.

livia-feedback

**Comment by livike** [comment]
Thanks @ercchy I'working right now on the improved version of the form. The actual is a simple {{form.as_p }} rendering in template and with default field type definitions.
Opening a FB topic for some help.

File: web/views.py (hunk 9 above)
Comment (livike) [inline web/views.py]: @ercchy : Pushed a change regarding this: stopped using 404, but .get() and also tried filter() are however raising an error. What I'm missing?

**Comment by ercchy** [comment]
:+1:

File: web/views.py (hunk 9 above)
Comment (goranche) [inline web/views.py]: > What I'm missing?

an error description, for starters...

Comment (ercchy) [inline web/views.py]: @livike I am not sure, bcs you said nothing about what kind of an error you get

Comment (livike) [inline web/views.py]: Ok. What I'm talking about is all about 404 error handling.
@ercchy wrote: 
"It is much better to get an empty dict (which the model will return if there will be not projects) and test for that in the template than this."
If testing on non existing project for example http://127.0.0.1:8000/project/11/edit/ It's raising an error (DoesNotExist at /project/11/edit/  Project matching query does not exist.) and not an empty dict. Probably I'm missing the point what Erika meant to said.
